File: src/aiteam/api/routes/analytics.py
"""AI Team OS — 数据分析统计路由."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query

from aiteam.api.deps import get_repository
from aiteam.storage.repository import StorageRepository

router = APIRouter(prefix="/api/analytics", tags=["analytics"])


@router.get("/tool-usage")
async def get_tool_usage(
    team_id: str | None = Query(None),
    repo: StorageRepository = Depends(get_repository),
) -> dict[str, Any]:
    """工具使用分布统计."""
    data = await repo.count_activities_by_tool(team_id=team_id)
    return {"success": True, "data": data}


@router.get("/agent-productivity")
async def get_agent_productivity(
    team_id: str | None = Query(None),
    repo: StorageRepository = Depends(get_repository),
) -> dict[str, Any]:
    """Agent产能指标."""
    data = await repo.get_agent_productivity(team_id=team_id)
    return {"success": True, "data": data}


@router.get("/timeline")
async def get_activity_timeline(
    team_id: str | None = Query(None),
    hours: int = Query(24, ge=1, le=168),
    repo: StorageRepository = Depends(get_repository),
) -> dict[str, Any]:
    """活动时间线（按小时聚合）."""
    data = await repo.get_activity_timeline(team_id=team_id, hours=hours)
    return {"success": True, "data": data}


@router.get("/efficiency")
async def get_efficiency_metrics(
    team_id: str | None = Query(None),
    repo: StorageRepository = Depends(get_repository),
) -> dict[str, Any]:
    """效率分析指标."""
    # 1. 任务完成率 + 平均完成时间
    task_stats = await repo.get_task_completion_stats(team_id=team_id)

    # 2. Agent利用率
    agent_utilization = await repo.get_agent_utilization(team_id=team_id)

    # 3. 工具调用效率：完成任务的平均工具调用次数
    productivity = await repo.get_agent_productivity(team_id=team_id)
    total_activities = sum(p["activity_count"] for p in productivity)
    completed = task_stats["completed_tasks"]
    avg_tools_per_task = (
        round(total_activities / completed, 2) if completed > 0 else None
    )

    # 4. 最高效Agent排行（按activities_per_hour降序）
    top_agents = sorted(
        agent_utilization,
        key=lambda a: a["activities_per_hour"],
        reverse=True,
    )[:5]

    return {
        "success": True,
        "data": {
            "task_completion": task_stats,
            "avg_tools_per_task": avg_tools_per_task,
            "agent_utilization": agent_utilization,
            "top_agents": top_agents,
        },
    }


@router.get("/team-overview")
async def get_team_overview(
    team_id: str = Query(...),
    repo: StorageRepository = Depends(get_repository),
) -> dict[str, Any]:
    """团队整体统计概览."""
    # 工具分布
    tool_dist = await repo.count_activities_by_tool(team_id=team_id)

    # Agent产能
    productivity = await repo.get_agent_productivity(team_id=team_id)

    # 活跃Agent数
    agents = await repo.list_agents(team_id)
    active_agents = [a for a in agents if a.status.value == "busy"]

    # 总活动数
    total_activities = sum(p["activity_count"] for p in productivity)

    return {
        "success": True,
        "data": {
            "total_activities": total_activities,
            "total_agents": len(agents),
            "active_agents": len(active_agents),
            "tool_distribution": tool_dist,
            "agent_productivity": productivity,
        },
    }

File: dashboard/src/api/analytics.ts
import { useQuery } from '@tanstack/react-query';
import { apiFetch } from './client';

export interface ToolUsageItem {
  tool_name: string;
  count: number;
}

export interface AgentProductivityItem {
  agent_id: string;
  agent_name: string;
  activity_count: number;
  tools_used: number;
  last_active: string;
}

export interface TimelineItem {
  hour: string;
  count: number;
}

export interface TeamOverview {
  total_activities: number;
  total_agents: number;
  active_agents: number;
  tool_distribution: ToolUsageItem[];
  agent_productivity: AgentProductivityItem[];
}

interface ApiResponse<T> {
  success: boolean;
  data: T;
}

function analyticsQuery(teamId?: string, projectId?: string, hours?: number): string {
  const params = new URLSearchParams();
  if (teamId) params.set('team_id', teamId);
  if (projectId) params.set('project_id', projectId);
  if (hours !== undefined) params.set('hours', String(hours));
  return params.size ? `?${params}` : '';
}

export function useToolUsage(teamId?: string, projectId?: string) {
  return useQuery({
    queryKey: ['analytics', 'tool-usage', teamId, projectId],
    queryFn: async () => {
      const res = await apiFetch<ApiResponse<ToolUsageItem[]>>(
        `/api/analytics/tool-usage${analyticsQuery(teamId, projectId)}`,
      );
      return res.data;
    },
    refetchInterval: 30_000,
  });
}

export function useAgentProductivity(teamId?: string, projectId?: string) {
  return useQuery({
    queryKey: ['analytics', 'agent-productivity', teamId, projectId],
    queryFn: async () => {
      const res = await apiFetch<ApiResponse<AgentProductivityItem[]>>(
        `/api/analytics/agent-productivity${analyticsQuery(teamId, projectId)}`,
      );
      return res.data;
    },
    refetchInterval: 30_000,
  });
}

export function useActivityTimeline(teamId?: string, hours = 24, projectId?: string) {
  return useQuery({
    queryKey: ['analytics', 'timeline', teamId, hours, projectId],
    queryFn: async () => {
      const res = await apiFetch<ApiResponse<TimelineItem[]>>(
        `/api/analytics/timeline${analyticsQuery(teamId, projectId, hours)}`,
      );
      return res.data;
    },
    refetchInterval: 60_000,
  });
}

export interface TaskCompletion {
  total_tasks: number;
  completed_tasks: number;
  completion_rate: number;
  avg_completion_hours: number | null;
}

export interface AgentUtilizationItem {
  agent_id: string;
  agent_name: string;
  activity_count: number;
  tools_used: number;
  span_hours: number;
  activities_per_hour: number;
  first_active: string | null;
  last_active: string | null;
}

export interface EfficiencyMetrics {
  task_completion: TaskCompletion;
  avg_tools_per_task: number | null;
  agent_utilization: AgentUtilizationItem[];
  top_agents: AgentUtilizationItem[];
}

export function useEfficiencyMetrics(teamId?: string, projectId?: string) {
  return useQuery({
    queryKey: ['analytics', 'efficiency', teamId, projectId],
    queryFn: async () => {
      const res = await apiFetch<ApiResponse<EfficiencyMetrics>>(
        `/api/analytics/efficiency${analyticsQuery(teamId, projectId)}`,
      );
      return res.data;
    },
    refetchInterval: 30_000,
  });
}

export function useTeamOverview(teamId: string) {
  return useQuery({
    queryKey: ['analytics', 'team-overview', teamId],
    queryFn: async () => {
      const res = await apiFetch<ApiResponse<TeamOverview>>(
        `/api/analytics/team-overview?team_id=${teamId}`,
      );
      return res.data;
    },
    enabled: !!teamId,
    refetchInterval: 30_000,
  });
}
